Speed up checks with caching and build matrices

jpculp opened the pull request:
*Description of changes:*

This commit refactors the GitHub Actions workflow by taking advantage of
build matrices and cached artifacts.

The cache is keyed off of Cargo.lock, but includes a partial key restore
so a cache is still used even when Cargo.lock changes. It will then
create a new cache for future builds of the same Cargo.lock.

A file named cache_bust is added to the .github folder. Any edit to that
file will cause a complete cache miss. We can use this to start over if
we ever run into trouble.

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

jpculp commented:
This is an alternative to my other PR (#76) which only cached the `cargo-deny` binary.

jpculp commented:
**Minimum actions duration:**
Before adding the license check: ~ 3m 46s
After adding the license check: ~ 6m 22s
This PR when utilizing fully cached artifacts: ~ 38s

tjkirch commented on .github/workflows/rust.yml:
```diff
@@ -3,33 +3,53 @@ on:
-      - run: cargo build --features rusoto-native-tls --no-default-features --locked
-      - run: cargo test --features rusoto-native-tls --no-default-features --locked
-  check-licenses:
+    - run: echo "CARGO_HOME=${{ github.workspace }}/.cargo" >> $GITHUB_ENV
```
Do we need this?  I believe GITHUB_ENV is for cases where you have to dynamically determine the variable name/contents, but we know it's always `CARGO_HOME=.cargo` (working directory should always be the same) so I think we can set it with a normal `env:` declaration in the YAML, no?

jpculp replied:
I was seeing issues with the `- run: rustup update stable` when **CARGO_HOME** was set to something else. The step above lets us set **CARGO_HOME** after the update step completes. If we remove that update step, then we can set via `env:`.

tjkirch commented on .github/workflows/rust.yml:
```diff
@@ -3,33 +3,53 @@ on:
+          ${{ hashFiles('.github/cache_bust') }}-${{ runner.os }}-${{ matrix.id }}
     - run: rustup update stable
-    - run: cargo install --version 0.6.2 cargo-deny --no-default-features
-    - run: cargo deny check --disable-fetch licenses
```
It seems like we're losing this actual call to cargo deny?

jpculp replied:
This did get lost, but was re-added in this latest push.

tjkirch commented on .github/workflows/rust.yml:
```diff
@@ -3,33 +3,53 @@ on:
+    - run: echo "CARGO_HOME=${{ github.workspace }}/.cargo" >> $GITHUB_ENV
+    - run: cargo build --features ${{ matrix.features }} ${{ matrix.additional_flags }} --locked
+    - run: cargo test --features ${{ matrix.features }} ${{ matrix.additional_flags }} --locked
+  check:
```
I'm not sure I understand the benefit of separating out the check job; it makes the workflow harder to understand, and we know the fmt/clippy steps will run faster right after a build, and we know the cargo-deny install will be fast with caching; can we run all the steps in one job instead?  fmt/clippy will be repeated, but I think that's correct - since we use features, the code they're checking is actually different each time.  Same with deny, the licenses might be different because of different feature dependencies.

jpculp committed:
speed up checks with caching and build matrices

This commit refactors the GitHub Actions workflow by taking advantage of
build matrices and cached artifacts.

The cache is keyed off of Cargo.lock, but includes a partial key restore
so a cache is still used even when Cargo.lock changes. It will then
create a new cache for future builds of the same Cargo.lock.

A file named cache_bust is added to the .github folder. Any edit to that
file will cause a complete cache miss. We can use this to start over if
we ever run into trouble.

jpculp commented on .github/workflows/rust.yml:
```diff
@@ -1,35 +1,40 @@
   pull_request:
     branches: [develop]
+    paths-ignore:
+      - '**.md'
```
This should skip the workflow if someone is just making a markdown change in a PR. The workflow will still occur on merge, unless we want it to be skipped there too.

tjkirch replied:
Would a change to a markdown file ever affect the cache?  Seems like we wouldn't need it.

jpculp replied:
I can't imagine how it would impact a build, so I'm not opposed to adding it. If we have it check on push it would simply be out of an extreme abundance of caution.

jpculp commented:
This latest push is a simplified version that is a little less parallel, but easier to follow. The first time it generates cached artifacts is a bit slower than before, but subsequent PR and push workflows should be very quick.

First iteration of PR when utilizing fully cached artifacts: ~ 38s
This simplified iteration when utilizing fully cached artifacts: ~ 50s

jpculp commented:
Added the logic to skip workflow on markdown changes for push (merge) as well.

jpculp commented:
Rebased.

tjkirch commented on .github/workflows/rust.yml:
```diff
@@ -1,35 +1,42 @@
+          .cargo
+          target
+        # you can edit the .github/cache_bust file if you need to clear the cache
+        key: ${{ hashFiles('.github/cache_bust') }}-${{ hashFiles('.github/workflows/rust.yml') }}-${{ runner.os }}-${{ matrix.features }}-${{ hashFiles('**/Cargo.lock') }}
```
I think we'd want to include additional_flags in the key and restore-keys, right?  There could be a change to flags that affects builds.

jpculp replied:
I'm not sure it would make much difference. I think in that scenario the hashes would be different and the cached bits with the wrong hashes would be ignored. My understanding of what is being accomplished here is we have 3 separate caches (one for each type of build) to prevent clobbering/cross-contamination. If we were to add additional_flags into the mix we'd just take up additional storage space by storing 3+_n_ caches.
